### DISCUSSION TITLE
Supporting mutiple `State` selectors without Substates

### DISCUSSION THREAD

**Discussion by auyer** [tokio-rs/axum#2220] [Ideas]
Hi

When creating a handler, we can use different `State` selectors.
```rust
pub async fn remove_key(
    Path(key): Path<String>,
    State(kvstore): State<KVStore>,
    State(actions): State<ActionBroadcaster>,
)...
```
But when creating the routes, we need to encapsulate them into a single structure.
```rust
// the application state that encapsulates both KVStore and ActionBroadcaster
#[derive(Clone)]
struct AppState {
    kv_store: KVStore,
    actions: ActionBroadcaster,
}

impl FromRef<AppState> for KVStore {
...
```

Would it be possible/desirable to just support using `.with_state(kv_store).with_state(actions)` instead of creating a new struct ?
If possible and desirable I would like to contribute this to the project.

Thanks!

**Comment by jplatte** [comment]
No, this is not possible with the state abstraction as-is. I think it would be possible to build something like that (using a type list), but I think it might require removing substates altogether, and also it would just be incredibly complex and possibly hurt compile times and inference. I don't think it's something that makes sense for axum.

**Reply by auyer** [reply]
Thanks !
Looking forward to contributing in other fronts in the future.
